content/en/blog/_posts/2025-01-14-devices-failure-handling/index.md

```diff
@@ -0,0 +1,403 @@
+significantly impact performance and lead to frustrating interruptions. As
+highlighted in the recent [Llama
+paper](https://ai.meta.com/research/publications/the-llama-3-herd-of-models/),
+hardware issues, particularly GPU failures, are a major cause of disruption in
+AI/ML training. You can also learn how much effort NVIDIA spend on handing
+devices failures and maintenance in the KubeCon talk by [Ryan Hallisey and Piotr
+Prokop All-Your-GPUs-Are-Belong-to-Us: An Inside Look at NVIDIA's Self-Healing
+GeForce NOW
+Infrastructure](https://kccncna2024.sched.com/event/1i7kJ/all-your-gpus-are-belong-to-us-an-inside-look-at-nvidias-self-healing-geforce-now-infrastructure-ryan-hallisey-piotr-prokop-pl-nvidia)
+([recording](https://www.youtube.com/watch?v=iLnHtKwmu2I)) as they see 19
+remediation requests per 1000 nodes a day!
+
+We also see data centers offering spot consumption models and overcommit on
+power, making device failures commonplace and a part of the business model.
+However Kubernetes’s view on resources is still very static. The resource is
+either there or not. And if it is there, the assumption is that it will stay
+there. This assumption combined with the overall complexity of a set up, leads
+to a variety of failure modes, which we discuss here.
+
+### Understanding AI/ML workloads
+
+Generally, all AI/ML workloads need specialized hardware, have challenging
+scheduling requirements, and are expensive when idle. AI/ML workloads typically
+fall into two categories - training and inference. Here is an oversimplified
+view on those categories’ characteristics, which are different traditional,
+workloads like web services:
+
+* **Training:** These workloads are resource-intensive, often consuming entire
+  machines and running as gangs of pods. Training jobs are usually "run to
+  completion" - but that could be days, weeks or even months. Any failure in a
+  single pod can necessitate restarting the entire step across all the pods.  
+* **Inference:** These workloads are usually long-running or run indefinitely,
+  and can be small enough to consume a subset of a Node’s devices or large to
+  span multiple nodes. They often require downloading huge files with the model
+  weights.
+
+These workload types specifically break many past assumptions:
+
+{{< table caption="Workload assumptions before and now" >}}
+| Before | Now |
+| :---- | :---- |
+| Can get a better CPU and the app will work faster. | Require a **specific** device (or **class of devices**) to run. |
+| When something doesn’t work, just recreate it. | Allocation or reallocation is expensive. |
+| Any node will work. No need to coordinate between Pods. | Scheduled in a special way - devices often connected in a cross nodes topology. |
+| Each Pod can be plug-and-play replaced if failed. | Pods are a part of a larger "task". Lifecycle of an entire "task" depends on each Pod. |
+| Images are slim and easily available. | Images are big and require special handling. |
```

Comment:
```suggestion
| Container images are slim and easily available. | Container images may be so big that they require special handling. |
```
You don't have to put your weights, etc into the image. But you still might want to.